gh-112075: Make instance attributes stored in inline "dict" thread safe

DinoV opened the pull request:
Adds some locking around when we de-materialize a dictionary.  

Also adds a comment where we'll need qsbr support for materialization.

<!-- gh-issue-number: gh-112075 -->
* Issue: gh-112075
<!-- /gh-issue-number -->

colesbury commented on Objects/dictobject.c:
```diff
@@ -5572,19 +5576,26 @@ _PyObject_MakeInstanceAttributesFromDict(PyObject *obj, PyDictOrValues *dorv)
     }
     assert(_PyType_HasFeature(Py_TYPE(obj), Py_TPFLAGS_HEAPTYPE));
+    bool success = true;
+    Py_BEGIN_CRITICAL_SECTION(dict);
```
I don't think we want a critical section on the dict here.

The dematerialization is new so I'm not entirely sure how we'll want to deal with it. The `nogil-3.12` fork had an assumption that materialization was at most one time per-object, which is no longer the case.

I think the thread-safety issues here are going to be around:
1) The refcount -- we need to ensure that we hold the only reference to the dict
2) The dict-or-values-pointer. Some other thread may access or try to modify it

But not so much the dict itself (since we need to ensure that we hold the only reference)

colesbury commented on Objects/dictobject.c:
```diff
@@ -6005,6 +6005,10 @@ has_unique_reference(PyObject *op)
 {
+#ifdef Py_GIL_DISABLED
+    // We don't yet support dematerialization in no-GIL builds
+    return false;
```
Since we are disabling this in the free-threaded build, we should get rid of `has_unique_reference` to fix the unused function warning and replace it with a simple `Py_REFCNT(dict) != 1`

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -655,24 +656,73 @@ _PyDictOrValues_IsValues(PyDictOrValues dorv)
 _PyDictOrValues_GetDict(PyDictOrValues dorv)
 {
     assert(!_PyDictOrValues_IsValues(dorv));
+#ifdef Py_GIL_DISABLED
```
Matt added some wrappers for this pattern to `pycore_atomic_ft_wrappers.h`

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -655,24 +656,73 @@ _PyDictOrValues_IsValues(PyDictOrValues dorv)
 }
 
+static inline PyDictValues *
+_PyDictOrValues_TryGetValues(PyObject *obj)
```
I think this function needs to be split up for read vs. write cases. The current logic doesn't seem quite right to me.

For readers, we generally want to fall back to a slower locking code path if it's not owned by the current thread and not marked shared.

For writers -- or really any case that hold's the lock -- we don't want a nested critical section. We also don't require that the object be marked as shared, although writers may mark it as shared as an optimization (so that future readers from non-owning threads don't need to acquire the lock).

DinoV replied:
Ahh yes, I'm missing locking on writes entirely...  I was thinking maybe we wouldn't need that for the shared values but that's not right.

I don't think we want to fall back to a slower locked version for reads here, mainly because it complicates the changes down stream in a similar way to making the locks for writes be present.  We'd need to have something like a TryGetValue instead of TryGetValues and then all of the consumers would need to pass in the index.  Instead we just need to lock to make the object shared and we can re-run the unlocked version safely.

colesbury replied:
> I don't think we want to fall back to a slower locked version for reads here, mainly because it complicates the changes down stream in a similar way to making the locks for writes be present.

That seems reasonable. I'm a bit worried that the more complex structure of `_PyDictOrValues_TryGetValues` will prevent efficient inlining, but we can address that later.

colesbury commented on Include/internal/pycore_dict.h:
```diff
@@ -264,6 +264,37 @@ _PyDictValues_AddToInsertionOrder(PyDictValues *values, Py_ssize_t ix)
+
+    PyDictValues *values = _PyDictOrValues_GetValues(*dorv_ptr);
+    PyObject *old_value = _Py_atomic_load_ptr_relaxed(&values->values[index]);
+    _Py_atomic_store_ptr_relaxed(&values->values[index], value);
```
Should be "release". For objects/pointers that are accessed without locks, I think we generally want at least "release" for stores to ensure that any writes in the pointed-to data are visible before the pointer.

colesbury commented on Include/internal/pycore_dict.h:
```diff
@@ -264,6 +264,37 @@ _PyDictValues_AddToInsertionOrder(PyDictValues *values, Py_ssize_t ix)
+        goto exit;
+    }
+
+    PyDictValues *values = _PyDictOrValues_GetValues(*dorv_ptr);
```
~~I think it'd be better if the check and the retrieval of the values were combined instead of having multiple loads of the values array.~~

EDIT: I think this needs a bit of restructuring. The `*dorv_ptr` is the pointer load that probably needs to be atomic.

DinoV replied:
With fixing the atomic loads the compiler should be able to optimize this down to a single read.  This is basically what the pattern typically was before this change and adding `_PyDictOrValues_TryGetValues` which I don't want to use here because we're already locking.

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -652,27 +654,64 @@ _PyObject_DictOrValuesPointer(PyObject *obj)
+static inline void
+_PyDictOrValues_SetDict(PyDictOrValues *ptr, PyObject *dict)
+{
+    FT_ATOMIC_STORE_PTR_RELAXED(ptr->dict, dict);
```
Should be "release"

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -652,27 +654,64 @@ _PyObject_DictOrValuesPointer(PyObject *obj)
 _PyDictOrValues_IsValues(PyDictOrValues dorv)
 {
-    return ((uintptr_t)dorv.values) & 1;
+    return ((uintptr_t)FT_ATOMIC_LOAD_PTR_RELAXED(dorv.values)) & 1;
```
I'm a bit confused by this: the previous code did not have a pointer load. I think this is doing an atomic load from a new pointer to a local variable -- that doesn't look right to me.

The atomic load should be in the caller to this function. By the time we get to this function we have a value.

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -652,27 +654,64 @@ _PyObject_DictOrValuesPointer(PyObject *obj)
 {
     assert(_PyDictOrValues_IsValues(dorv));
-    return (PyDictValues *)(dorv.values + 1);
+    return (PyDictValues *)(((char *)FT_ATOMIC_LOAD_PTR_RELAXED(dorv.values)) + 1);
```
Same here. I don't think this atomic load is right.

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -652,27 +654,64 @@ _PyObject_DictOrValuesPointer(PyObject *obj)
 {
     assert(!_PyDictOrValues_IsValues(dorv));
-    return dorv.dict;
+    return FT_ATOMIC_LOAD_PTR_RELAXED(dorv.dict);
```
Same here. I don't think this atomic load is right.

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -652,27 +654,64 @@ _PyObject_DictOrValuesPointer(PyObject *obj)
 _PyDictOrValues_SetValues(PyDictOrValues *ptr, PyDictValues *values)
 {
-    ptr->values = ((char *)values) - 1;
+    FT_ATOMIC_STORE_PTR_RELAXED(ptr->values, (char *)values - 1);
```
Should be "release"

colesbury commented on Objects/dictobject.c:
```diff
@@ -6686,21 +6718,15 @@ PyObject_GenericGetDict(PyObject *obj, void *context)
     if (_PyType_HasFeature(tp, Py_TPFLAGS_MANAGED_DICT)) {
         PyDictOrValues *dorv_ptr = _PyObject_DictOrValuesPointer(obj);
         if (_PyDictOrValues_IsValues(*dorv_ptr)) {
```
Regarding the atomics in `_PyDictOrValues_IsValues`. I think we need the `*dorv_ptr` access to be atomic -- not the accesses within `_PyDictOrValues_IsValues`.

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -645,6 +646,20 @@ _PyObject_ManagedDictPointer(PyObject *obj)
+_PyObject_SetManagedDict(PyObject *obj, PyObject *dict)
+{
+    PyManagedDictPointer *ptr = _PyObject_ManagedDictPointer(obj);
+    FT_ATOMIC_STORE_PTR_RELEASE(ptr->dict, dict);
```
Needs a cast to fix the warning

colesbury commented on Objects/dictobject.c:
```diff
@@ -6723,6 +6736,60 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+    // If we haven't materialized the dictionary yet we lock on the object, which
+    // will also be used to prevent the dictionary from being materialized while
+    // we're doing the insertion.  If we race and the dictionary gets created
+    // then we'll need to relace the object lock and lock the dictionary to
```
relace -> release

colesbury commented on Objects/dictobject.c:
```diff
@@ -6723,6 +6736,60 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+
+        if (dict == NULL) {
+            res = store_instance_attr_lock_held(obj, values, name, value);
+        } else {
```
`else` on new line

colesbury commented on Objects/dictobject.c:
```diff
@@ -6723,6 +6736,60 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+        if (retry_with_dict) {
+            goto with_dict;
+        }
+    } else {
```
`else` on new line

colesbury commented on Objects/dictobject.c:
```diff
@@ -6723,6 +6736,60 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+        Py_BEGIN_CRITICAL_SECTION(dict);
+        if (dict->ma_values == values) {
+            res = store_instance_attr_lock_held(obj, values, name, value);
+        } else {
```
`else` on new line

colesbury commented on Objects/dictobject.c:
```diff
@@ -6754,19 +6821,52 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
+        int res = 0;
+        // Try again with the object locked...
+        Py_BEGIN_CRITICAL_SECTION(obj);
+        // We could actually race with a dictioanry resize here.  The
```
dictioanry -> dictionary

colesbury commented on Objects/dictobject.c:
```diff
@@ -6754,19 +6821,52 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
+        // values, so if we're racing with a resize + insert into the
+        // value we're reading it's just like we won the race.
+        if (!values->valid) {
+            res = 1;
```
I don't understand this case:
1) I don't understand the comment in relation to the code
2) It looks like this will cause `_PyObject_GenericGetAttrWithDict` to return `NULL` without an error set

DinoV replied:
This is related to your comment below about this not being thread-safe, i.e. the concurrent modification to the dictionary which could be making the values invalid without `obj` being locked.  I was thinking that even if that happened we'd still have the old values in the concurrently becoming invalid inline values.  But that could also be getting decref'd which would definitely be an issue.

I don't think we'll return NULL w/o an error set?  The return value of 1 here indicates that the caller needs to re-try the lookup from the dictionary as the values are definitely no longer valid.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6754,19 +6821,52 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
+        }
+
+        value = _Py_atomic_load_ptr_relaxed(&values->values[ix]);
+        *attr = Py_XNewRef(value);
```
Is this safe? I think the values can be modified concurrently through the dictionary (without locking `obj`)

DinoV replied:
You're right, it's not...  I was hoping we could avoid the double lock dance here but I do think we need it.

DinoV committed:
Make instance attributes stored in inline "dict" thread safe

colesbury commented on Objects/dictobject.c:
```diff
@@ -6681,25 +6708,21 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
         if (dict == NULL) {
-            dict = make_dict_from_instance_attributes(
-                    interp, keys, values);
+            dict = _PyObject_MaterializeManagedDict(obj);
```
We might already be holding the object locks. Not sure if that's common enough to be an issue.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6681,25 +6708,21 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
-            return PyDict_SetItem((PyObject *)dict, name, value);
         }
+        // Caller needs to insert into materialized dict
+        return 1;
```
Why can't we insert/delete from the dict here like before? Is it because we are holding the object lock and would need to hold the dict lock instead?

DinoV replied:
Yep, that is a problem...  We can know what state we're in though and can lock appropriately.  I don't know that I like quite how it turns out, but let me know what you think about the latest version.  I can always go back to the tri-return value and documenting it.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6624,27 +6624,54 @@ make_dict_from_instance_attributes(PyInterpreterState *interp,
 
-int
-_PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+static int
+store_instance_attr_lock_held(PyObject *obj, PyDictValues *values,
```
I was a bit confused by two things:

1) The return value can be `-1`, `0`, or `1` where `1` means "retry with materialized managed dict". It'd be nice to avoid that, but if that's not practical then let's document the meanings of the return values at the function level.
2) This is "lock_held", but sometimes it's the object's lock and sometimes it's the dict's lock. I'd love for that to be simpler, but if that's not possible let's also document that.

DinoV replied:
I'll change it to do the insert here.  I was originally just keeping the caller of `_PyObject_TryStoreInstanceAttribute` the same where it's checking if the values are still valid, but there's no reason why we can't just handle the store to the dict.

I would love for it to be simpler, the double lock dance feels really ugly, but I'm not sure how we can handle it otherwise.  We could do something like get back to the object with a function which does the inverse of `_PyObject_InlineValues` to lock the object from the dict, but I think we'd still need to lock the dict to not race with detaching, so it doesn't help much (and might just complicate all of the mutations that can come from the dictionary APIs).  So I'll add a comment.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6757,19 +6837,80 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
+
+#ifdef Py_GIL_DISABLED
+    PyObject *value = _Py_atomic_load_ptr_relaxed(&values->values[ix]);
+    if (value == NULL || _Py_TryIncref(&values->values[ix], value)) {
```
Should this be `value != NULL`?

DinoV replied:
Nope, this is a case where the value hasn't been assigned yet.  The attribute can exist in the shared keys but not exist in the instance.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6757,19 +6837,80 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
+        if (dict == NULL) {
+            // Still no dict, we can read from the values
+            assert(values->valid);
+            value = _Py_atomic_load_ptr_relaxed(&values->values[ix]);
```
Does this need to be atomic? Can anyone else be writing to values if we are holding the object's lock?

DinoV replied:
No, I think this is left over from when store instance wasn't always locking.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6827,38 +6967,95 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
 int
 _PyDict_DetachFromObject(PyDictObject *mp, PyObject *obj)
 {
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(mp);
```
This also has a `Py_BEGIN_CRITICAL_SECTION(mp)` in the function body.

DinoV committed:
Insert into dict on store if dict is materialized and other feedback

colesbury commented on Objects/dictobject.c:
```diff
@@ -6615,28 +6615,63 @@ make_dict_from_instance_attributes(PyInterpreterState *interp,
-    return make_dict_from_instance_attributes(interp, keys, values);
+    PyDictObject *dict = make_dict_from_instance_attributes(interp, keys, values);
+    FT_ATOMIC_STORE_PTR_RELEASE(_PyObject_ManagedDictPointer(obj)->dict,
+                            (PyDictObject *)dict);
```
formatting: align `(PyDictObject *)dict`

colesbury commented on Objects/dictobject.c:
```diff
@@ -6749,19 +6869,80 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
+
+#ifdef Py_GIL_DISABLED
+    PyObject *value = _Py_atomic_load_ptr_relaxed(&values->values[ix]);
+    if (value == NULL || _Py_TryIncref(&values->values[ix], value)) {
```
`s/_Py_TryIncref/_Py_TryIncrefCompare` now

colesbury commented on Objects/object.c:
```diff
@@ -1473,10 +1473,9 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
-        PyObject *attr = _PyObject_GetInstanceAttribute(obj, values, name);
+    PyObject *dict, *attr;
+    if ((tp->tp_flags & Py_TPFLAGS_INLINE_VALUES) &&
+         !_PyObject_TryGetInstanceAttribute(obj, name, &attr)) {
```
I was confused that `!_PyObject_TryGetInstanceAttribute` means that the attribute was successfully retrieved.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6718,6 +6773,71 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+        else {
+            // We lost a race with the materialization of the dict, we'll
+            // try the insert with it...
+            goto with_dict;
```
This jumps out of the critical section without unlocking the lock.

Let's also avoid gotos jumping inside blocks (i.e., jumping to inside the else statement).

colesbury commented on Objects/dictobject.c:
```diff
@@ -6819,38 +6999,95 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
     assert(_PyObject_InlineValuesConsistencyCheck(obj));
     PyTypeObject *tp = Py_TYPE(obj);
     if (tp->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
         PyDictObject *dict = _PyObject_ManagedDictPointer(obj)->dict;
```
Does this need to be `_PyObject_GetManagedDict`?

colesbury commented on Objects/dictobject.c:
```diff
@@ -6876,29 +7114,28 @@ PyObject_GenericGetDict(PyObject *obj, void *context)
-            if (dict != NULL) {
-                managed_dict->dict = (PyDictObject *)dict;
-            }
+            _PyObject_InlineValues(obj)->valid) {
```
Does this need an atomic load of `valid`?

colesbury commented on Objects/dictobject.c:
```diff
@@ -6819,38 +6999,95 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+{
+    PyDictValues *values = _PyObject_InlineValues(obj);
+
+    if (values->valid) {
```
The order of operations here seems strange. I think it means that concurrent reads can see empty or partially emptying dicts.

Can we assign the new dict, then set `valid` to 0, and then clear the old values?

colesbury commented on Objects/typeobject.c:
```diff
@@ -6190,15 +6190,17 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
+
+            Py_BEGIN_CRITICAL_SECTION2(self, dict);
+
+            if (dict == NULL || _PyDict_DetachFromObject(dict, self)) {
```
suggestion: `_PyDict_DetachFromObject(dict, self) < 0`

colesbury commented on Objects/object.c:
```diff
@@ -1699,12 +1694,13 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
-                    obj, _PyObject_InlineValues(obj), name, value);
+
+        if ((tp->tp_flags & Py_TPFLAGS_INLINE_VALUES)) {
+            res = _PyObject_TryStoreInstanceAttribute(obj, name, value);
```
Why does this have the `Try` prefix? Can it fail in ways that a "normal" function like `PyDict_SetItem` would not fail?

DinoV replied:
Nope, this is a hold over from the previous version where it returned a value indicating the caller needed to do the store to the dict directly.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6673,25 +6708,45 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
-                    interp, keys, values);
+            // Make the dict but don't publish it in the object
+            // so that no one else will see it.
+            dict = materialize_managed_dict_lock_held(obj);
```
I am confused by the comment. Doesn't `materialize_managed_dict_lock_held` internally "publish" the dict?

DinoV replied:
Indeed, this is just an outdated comment from a WIP when I was initially thinking that was the way to go.  I'll delete it.

DinoV committed:
_PyObject_TryStoreInstanceAttribute -> _PyObject_StoreInstanceAttribute
Fix issue where critical section isn't released
Make _PyObject_TryGetInstanceAttribute return a bool

colesbury commented on Objects/dictobject.c:
```diff
@@ -6819,38 +6993,98 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+
 void
-PyObject_ClearManagedDict(PyObject *obj)
+_PyObject_SetManagedDict(PyObject *obj, PyObject *new_dict)
```
This function (and related helper functions) could use some clean-up:

- Let's avoid the weird `goto` if possible
- The stores to `_PyObject_ManagedDictPointer(obj)->dict` need to be atomic
- I think it'll be clearer if you lift the fallible check in `set_dict_inline_values` into this function
- The assertions in `_PyDict_DetachFromObject` may not hold: once you lock the dict and object, the dict may no longer be attached.
- We need to assign `new_dict` before we mark the values as invalid

colesbury commented on Objects/dictobject.c:
```diff
@@ -6718,6 +6771,66 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+    PyDictValues *values = _PyObject_InlineValues(obj);
+    if (!FT_ATOMIC_LOAD_UINT8_RELAXED(values->valid)) {
+        return store_instance_attr_dict(obj, _PyObject_GetManagedDict(obj), name, value);
+    }
```
`_PyObject_GetManagedDict(obj)` can be `NULL` here:

The following crashes (we should add it or something similar as a unit test):

```python
class Foo:
    pass

f = Foo()
del f.__dict__
f.a = 3
```

colesbury reviewed (commented):
We need to use seq-cst atomics (as opposed to relaxed) for reads and writes to `values->valid`. For example, in `_PyObject_StoreInstanceAttribute`, the `!values->valid` check must imply that managed dict is not embedded:

- When we detach the dict, we must first assign the new dict and then mark the values as not valid (using seq-cst atomics)

colesbury commented on Objects/typeobject.c:
```diff
@@ -6190,15 +6190,17 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
             }
-            if (_PyDict_DetachFromObject(dict, self)) {
+
+            Py_END_CRITICAL_SECTION2();
```
The `dict == NULL` check has to be outside of the critical section. `Py_BEGIN_CRITICAL_SECTION2` does not handle `NULL` arguments.

DinoV committed:
Cleanup _PyObject_SetManagedDict, use cst for inline valid flag, fix issue w/ deleted dict

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -659,10 +660,10 @@ extern PyObject* _PyType_GetDocFromInternalDoc(const char *, const char *);
-PyObject * _PyObject_GetInstanceAttribute(PyObject *obj, PyDictValues *values,
-                                        PyObject *name);
+extern int _PyObject_StoreInstanceAttribute(PyObject *obj,
+                                               PyObject *name, PyObject *value);
```
alignment of `name`

colesbury commented on Objects/dictobject.c:
```diff
@@ -6749,19 +6869,80 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
-                              PyObject *name)
+// Attempts to get an instance attribute from the inline values.  Returns 0 if
+// the lookup from the inline values was successful or 1 if the inline values
+// are no longer valid.  No error is set in either case.
```
Comment is out of date

colesbury commented on Objects/dictobject.c:
```diff
@@ -6673,25 +6708,43 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+                res = PyDict_DelItem((PyObject *)dict, name);
+            } else {
+                res = PyDict_SetItem((PyObject *)dict, name, value);
+            }
```
These calls reacquire the dict's critical section

DinoV replied:
This actually shouldn't.  The reason why the previous comment was there regarding the object not being published yet was because the materialization publishes the dict and therefore it needs to be locked (in this case only the object is locked because the dict didn't previously exist).  

I was originally hoping to avoid re-locking the dict here and publishing it after.  But that complexity didn't quite seem like it was worth it due the extra complexity of the del item case.

But I guess we're doing that in 2 spots already, so I've factored that into a helper and now we'll call that here and publish the materialized dict at the end.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6718,6 +6771,73 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+            res = PyDict_DelItem((PyObject *)dict, name);
+        } else {
+            res = PyDict_SetItem((PyObject *)dict, name, value);
+        }
```
These calls reacquire the dict's critical section

colesbury commented on Objects/dictobject.c:
```diff
@@ -6819,53 +6999,124 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+
+        Py_XINCREF(new_dict);
+        FT_ATOMIC_STORE_PTR(_PyObject_ManagedDictPointer(obj)->dict,
+                            (PyDictObject *)Py_XNewRef(new_dict));
```
Is the `Py_XINCREF(new_dict);` redundant with the `Py_XNewRef(new_dict)`?

colesbury commented on Objects/dictobject.c:
```diff
@@ -6819,53 +6999,124 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+        PyDictObject *cur_dict = _PyObject_ManagedDictPointer(obj)->dict;
+        assert(cur_dict == dict ||
+               (cur_dict->ma_values != _PyObject_InlineValues(obj) &&
+                !_PyObject_InlineValues(obj)->valid));
```
I think this comment and assertion, the one below, and the one in typeobject.c can all be combined and place in `_PyDict_DetachFromObject`.

DinoV replied:
I would have liked to have done that :). 

But we'd need to pass in `cur_dict` to make that work (presumably as an argument like `prev_dict`), because once we're in `_PyDict_DetachFromObject` we've already replaced the dict w/ `new_dict`.  

And having the 2 dicts passed in I think might just make the API more confusing, especially given that the only 2 call sites are rather different in that one is actually replacing the dict and the other isn't.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6819,53 +6999,124 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+                dict->ma_values != _PyObject_InlineValues(obj));
+
+        if (cur_dict == dict) {
+            _PyDict_DetachFromObject(dict, obj);
```
I don't think the `if` statement is necessary. We already handle this case within `_PyDict_DetachFromObject` with the `mp->ma_values != _PyObject_InlineValues(obj)` check.

DinoV replied:
It does have the added benefit of avoiding a warning on not using `cur_dict` in release builds but I'll remove it and wrap the assertion in an ifdef :)

colesbury reviewed (commented):
This looks pretty good. Just a few small comments and what looks like an extra incref.

DinoV committed:
Refactor del/set into set_or_del_lock_held, and avoid locking on newly materialized dict
Fix duplicate incref
Fix comment
Remove redundant if check on detach

colesbury commented on Objects/dictobject.c:
```diff
@@ -6820,53 +7000,119 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+        // the dict and unlocked the objects then it's definitely no longer
+        // inline and there's no need to detach it, we can just replace it.
+        // The call to _PyDict_DetachFromObject will be a nop.
+        PyDictObject *cur_dict = _PyObject_ManagedDictPointer(obj)->dict;
```
A follow-up on my previous comment: we need to reload the dict within the lock. Otherwise, I think the `Py_DECREF(dict);` will be incorrect.

For example, I think this should be:

```
dict = _PyObject_ManagedDictPointer(obj)->dict;
```

In the case where `dict` changed, it doesn't really matter which one we try to detach, but it matters which one we `Py_DECREF()`.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6820,53 +7000,119 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+        assert(cur_dict == dict ||
+               (cur_dict->ma_values != _PyObject_InlineValues(obj) &&
+                dict->ma_values != _PyObject_InlineValues(obj) &&
+                !_PyObject_InlineValues(obj)->valid));
```
I still think this assertion is more confusing than helpful and that we can get nearly all the value with an assertion in `_PyDict_DetachFromObject()`, but that's a stylistic judgement and ultimately up to you.

DinoV replied:
I think part of the reason why I liked this is that there is some distinct oddness of passing `dict` to `_PyDict_DetachFromObject` when `cur_dict` is the actual dict, but it's not locked.

I'm tweaking this so that we'll pass the actual dict, we'll do a relaxed atomic load of `ma_values`, and then only assert that the dict is locked when we're doing the detached.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6820,53 +7000,119 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(mp);
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(obj);
+
     if (mp->ma_values == NULL || mp->ma_values != _PyObject_InlineValues(obj)) {
```
Related to the assertion in `_PyObject_SetManagedDict` I would include some of that here and maybe a comment about that case:

```suggestion
    if (mp->ma_values != _PyObject_InlineValues(obj)) {
        assert(!_PyObject_InlineValues(obj)->valid);
```

(`_PyObject_InlineValues()` can never be `NULL`)

DinoV replied:
I think maybe the only version of the assertion that could be included here is:
```
    assert(dict->ma_values != _PyObject_InlineValues(obj) ||
            _PyObject_InlineValues(obj)->valid));
```

But that basically already exists as `assert(mp->ma_values->valid == 1);` occurs after the inline values check.  I'm going to simplify the assertion in `_PyObject_SetManagedDict` though.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6820,53 +7000,119 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
     assert(Py_TYPE(obj)->tp_flags & Py_TPFLAGS_INLINE_VALUES);
-    Py_BEGIN_CRITICAL_SECTION(mp);
+
     mp->ma_values = copy_values(mp->ma_values);
```
This predates this PR, but I think this should assign to a temporary instead of directly to `mp->ma_values`. Otherwise, we can leave the dict in an invalid state if `copy_values` runs out of memory.

colesbury commented on Objects/dictobject.c:
```diff
@@ -6820,53 +7000,119 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
+
+        Py_END_CRITICAL_SECTION2();
+
+        Py_DECREF(dict);
```
i.e., this `Py_DECREF(dict)` would be incorrect without the reloading. Also, I think `dict` might be `NULL` after reloading so we should handle that.

DinoV committed:
Handle dict replacement case better

colesbury reviewed (approved):
LGTM

DinoV committed:
Decref materialized dict on fail to set

DinoV committed:
Fix missing decref when storing and dict is detached but has been deleted
